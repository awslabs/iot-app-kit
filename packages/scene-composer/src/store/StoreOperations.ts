export type OperationType = 'INITIALIZE' | 'UPDATE_DOCUMENT' | 'UPDATE_EDITOR' | 'TRANSIENT';

export type SceneComposerEditorOperation =
  | 'resetEditorState'
  | 'setCameraControlsType'
  | 'setEditorConfig'
  | 'setSelectedSceneNodeRef'
  | 'setSubModelSelection'
  | 'setHighlightedSceneNodeRef'
  | 'setLoadingModelState'
  | 'setTransformControlMode'
  | 'setTransformControls'
  | 'setCameraTarget'
  | 'addMessages'
  | 'clearMessages'
  | 'setAddingWidget'
  | 'setCursorPosition'
  | 'setCursorLookAt'
  | 'setCursorVisible'
  | 'setCursorStyle'
  | 'setActiveCameraSettings'
  | 'setActiveCameraName'
  | 'setMainCameraObject';

export type SceneComposerDocumentOperation =
  | 'loadScene'
  | 'appendSceneNodeInternal'
  | 'updateSceneNodeInternal'
  | 'updateSceneNodeInternalTransient'
  | 'updateDocumentInternal'
  | 'removeSceneNode'
  | 'updateSceneRuleMapById'
  | 'removeSceneRuleMapById'
  | 'addComponentInternal'
  | 'updateComponentInternal'
  | 'removeComponent'
  | 'setSceneProperty'
  | 'clearTemplatizedDataBindings';

export type SceneComposerDataOperation = 'setDataInput' | 'setDataBindingTemplate';

export type SceneComposerViewOptionOperation = 'toggleMotionIndicatorVisibility' | 'setTagSettings';

export type SceneComposerOperation =
  | SceneComposerEditorOperation
  | SceneComposerDocumentOperation
  | SceneComposerViewOptionOperation
  | SceneComposerDataOperation;

export const SceneComposerOperationTypeMap: Record<SceneComposerOperation, OperationType> = {
  // loadScene is a speical operation as we want to clear the undo/redo state after this
  loadScene: 'INITIALIZE',

  appendSceneNodeInternal: 'UPDATE_DOCUMENT',
  updateSceneNodeInternal: 'UPDATE_DOCUMENT',
  updateSceneNodeInternalTransient: 'TRANSIENT',
  updateDocumentInternal: 'UPDATE_DOCUMENT',
  removeSceneNode: 'UPDATE_DOCUMENT',
  addComponentInternal: 'UPDATE_DOCUMENT',
  updateComponentInternal: 'UPDATE_DOCUMENT',
  removeComponent: 'UPDATE_DOCUMENT',
  updateSceneRuleMapById: 'UPDATE_DOCUMENT',
  removeSceneRuleMapById: 'UPDATE_DOCUMENT',
  setSceneProperty: 'UPDATE_DOCUMENT',
  clearTemplatizedDataBindings: 'UPDATE_DOCUMENT',

  resetEditorState: 'UPDATE_EDITOR',
  setCameraControlsType: 'UPDATE_EDITOR',
  setEditorConfig: 'UPDATE_EDITOR',
  setSelectedSceneNodeRef: 'UPDATE_EDITOR',
  setSubModelSelection: 'UPDATE_EDITOR',
  setHighlightedSceneNodeRef: 'UPDATE_EDITOR',
  setTransformControlMode: 'UPDATE_EDITOR',
  setTransformControls: 'UPDATE_EDITOR',
  setCameraTarget: 'UPDATE_EDITOR',
  addMessages: 'UPDATE_EDITOR',
  clearMessages: 'UPDATE_EDITOR',
  setAddingWidget: 'UPDATE_EDITOR',
  setCursorPosition: 'UPDATE_EDITOR',
  setCursorLookAt: 'UPDATE_EDITOR',
  setCursorVisible: 'UPDATE_EDITOR',
  setCursorStyle: 'UPDATE_EDITOR',
  setActiveCameraSettings: 'UPDATE_EDITOR',
  setActiveCameraName: 'UPDATE_EDITOR',
  setMainCameraObject: 'UPDATE_EDITOR',

  setLoadingModelState: 'TRANSIENT',
  setDataInput: 'TRANSIENT',
  setDataBindingTemplate: 'TRANSIENT',

  toggleMotionIndicatorVisibility: 'TRANSIENT',
  setTagSettings: 'TRANSIENT',
};
