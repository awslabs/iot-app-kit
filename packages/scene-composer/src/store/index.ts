export * from './StoreOperations';
export * from './Store';
