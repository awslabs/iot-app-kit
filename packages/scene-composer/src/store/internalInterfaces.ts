import { IErrorDetails } from '../common/errors';
import {
  SelectionChangedEventCallback,
  IAnchorComponent,
  ICameraComponent,
  IColorOverlayComponent,
  ILightComponent,
  IModelRefComponent,
  IMotionIndicatorComponent,
  IRuleBasedMap,
  IRuleStatement,
  ISceneComponent,
  ISceneDocument,
  ISceneNode,
  ITransform,
  ITransformConstraint,
  IValueDataBinding,
  IValueDataBindingProvider,
  OperationMode,
  ShowAssetBrowserCallback,
  URIModifier,
  KnownComponentType,
  WidgetClickEventCallback,
  ISubModelRefComponent,
  IDataOverlayComponent,
} from '../interfaces';
import { MapControls as MapControlsImpl, OrbitControls as OrbitControlsImpl } from '../three/OrbitControls';

export type ISerializationErrorDetails = IErrorDetails;

export interface IDeserializationResult {
  document?: ISceneDocumentInternal;
  errors: ISerializationErrorDetails[];
}

/******************************************************************************
 * Editor Data Model
 ******************************************************************************/

export enum DisplayMessageCategory {
  Error = 'Error',
  Warning = 'Warning',
  Info = 'Info',
}

export type CameraControlImpl = MapControlsImpl | OrbitControlsImpl;
export type TweenValueObject = { x: number; y: number; z: number };

export type CursorStyle = 'move' | 'edit';

export interface IDisplayMessage {
  category: DisplayMessageCategory;
  messageText: string;
  params?: Record<string, string>;
}

export interface IEditorConfig {
  operationMode?: OperationMode;
  uriModifier?: URIModifier;
  valueDataBindingProvider?: IValueDataBindingProvider;
  showAssetBrowserCallback?: ShowAssetBrowserCallback;
  onWidgetClick?: WidgetClickEventCallback;
  onSelectionChanged?: SelectionChangedEventCallback;
}

/******************************************************************************
 * Document Model
 ******************************************************************************/

export type ITransformInternal = ITransform;
export type IRuleBasedMapInternal = IRuleBasedMap;
export type IRuleStatementInternal = IRuleStatement;

export interface ISceneNodeInternal extends ISceneNode {
  // below fields are not nullable internally
  ref: string;
  name: string;
  transform: ITransformInternal;
  transformConstraint: ITransformConstraint;
  components: ISceneComponentInternal[];
  childRefs: string[];
  properties: Record<string, any>;
}

export interface ISceneDocumentInternal extends ISceneDocument {
  unit: string; // unit is not nullable internally, default to 'meter'
  ruleMap: Record<string, IRuleBasedMapInternal>;
  nodeMap: Record<string, ISceneNodeInternal>;
  componentNodeMap: { [type in KnownComponentType | string]?: Record<string, string[]> }; // record of node.ref to component.ref[]
}

export interface ISceneComponentInternal extends ISceneComponent {
  // unique id for the component for internal reference
  ref: string;
}

export interface IDataBoundSceneComponentInternal extends ISceneComponentInternal {
  valueDataBinding?: IValueDataBinding;
}

/******************************************************************************
 * Components
 ******************************************************************************/

export type ILightComponentInternal = ISceneComponentInternal & ILightComponent;

export type ICameraComponentInternal = ISceneComponentInternal & ICameraComponent;

export type IModelRefComponentInternal = ISceneComponentInternal & IModelRefComponent;

export type ISubModelRefComponentInternal = ISceneComponentInternal & ISubModelRefComponent;

export type IAnchorComponentInternal = IDataBoundSceneComponentInternal & IAnchorComponent;

export type IColorOverlayComponentInternal = IDataBoundSceneComponentInternal & IColorOverlayComponent;

export type IMotionIndicatorComponentInternal = ISceneComponentInternal & IMotionIndicatorComponent;

export type IDataOverlayComponentInternal = ISceneComponentInternal & IDataOverlayComponent;

/******************************************************************************
 * Type magic...
 ******************************************************************************/

export const isISceneNodeInternal = (item: ISceneNodeInternal | undefined): item is ISceneNodeInternal => {
  return !!item;
};

export const isISceneComponentInternal = (
  item: ISceneComponentInternal | undefined,
): item is ISceneComponentInternal => {
  return !!item;
};
