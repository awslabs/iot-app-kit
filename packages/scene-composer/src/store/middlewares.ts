import { State, StateCreator } from 'zustand';
import { produce, Draft } from 'immer';
import createVanilla, { GetState, SetState, StoreApi } from 'zustand/vanilla';

import DebugLogger from '../logger/DebugLogger';

import { SceneComposerOperationTypeMap } from './StoreOperations';
import { RootState } from './Store';

const LOG = new DebugLogger('stateStore');

/**
 * Log the update to the state store, the prev state and the updated state.
 */
export const log =
  <T extends State>(config: StateCreator<T>): StateCreator<T> =>
  (set, get, api) => {
    return config(
      (...args) => {
        LOG.verbose('old state', get());
        set(...args);
        LOG.verbose('new state', get());
      },
      get,
      api,
    );
  };

/**
 * Make nested state update simple.
 */
export const immer =
  <T extends State>(config: StateCreator<T>): StateCreator<T> =>
  (set, get, api) =>
    config(
      (partial, replace) => {
        const nextState = typeof partial === 'function' ? produce(partial as (state: Draft<T>) => T) : (partial as T);
        set(nextState, replace);
      },
      get,
      api,
    );

/**
 * Undo/Redo, inspired by zundo
 */
export type UndoState = Partial<
  Pick<UndoStoreState, 'undo' | 'redo' | 'clear'> & {
    getUndoState: () => UndoStoreState;
    undoStore: StoreApi<UndoStoreState>;
    // last operation that changed the parent state
    lastOperation?: string;
  }
>;
export interface UndoStoreState {
  prevStates: any[];
  futureStates: any[];
  undo: () => void;
  redo: () => void;
  clear: () => void;
  // handle on the parent store's setter
  setStore: Function;
  // handle on the parent store's getter
  getStore: Function;
}

function filterNoHistoryStates(newState: RootState, currentState: RootState) {
  return {
    ...newState,
    noHistoryStates: currentState.noHistoryStates,
  };
}

// factory to create undoStore. contains memory about past and future states and has methods to traverse states
export const createUndoStore = () => {
  return createVanilla<UndoStoreState>((set, get) => {
    return {
      prevStates: [],
      futureStates: [],
      undo: () => {
        const { prevStates, futureStates, setStore, getStore } = get();
        if (prevStates.length > 0) {
          futureStates.push(getStore());
          const prevState = filterNoHistoryStates(prevStates.pop(), getStore());
          setStore(prevState);
          // force updating the undo store to trigger subscription on the store
          set({ prevStates, futureStates });
        }
      },
      redo: () => {
        const { prevStates, futureStates, setStore, getStore } = get();
        if (futureStates.length > 0) {
          prevStates.push(getStore());
          const futureState = filterNoHistoryStates(futureStates.pop(), getStore());
          setStore(futureState);
          // force updating the undo store to trigger subscription on the store
          set({ prevStates, futureStates });
        }
      },
      clear: () => {
        set({ prevStates: [], futureStates: [] });
      },
      setStore: () => {},
      getStore: () => {},
    };
  });
};

// custom zustand middleware to get previous state
export const undoMiddleware =
  <TState extends UndoState>(config: StateCreator<TState>) =>
  (set: SetState<TState>, get: GetState<TState>, api: StoreApi<TState>) => {
    const undoStore = createUndoStore();
    const { getState, setState } = undoStore;
    const { undo, clear, redo } = getState();
    return config(
      (args) => {
        // Take a snapshot of the current state
        const snapshot = {
          prevStates: [...getState().prevStates, { ...get() }],
          setStore: set,
          futureStates: [],
          getStore: get,
        };

        if (!get().getUndoState) {
          // inject helper functions to user defined store.
          set({
            undo,
            clear,
            redo,
            getUndoState: getState,
            undoStore,
          });
        }

        set(args);

        const lastOperation = get().lastOperation || '';
        const lastOperationType = SceneComposerOperationTypeMap[lastOperation];

        LOG.verbose('lastOperation', lastOperation, 'maps to', lastOperationType);

        if (lastOperationType === 'INITIALIZE') {
          // clear history
          LOG.verbose('clear all history after initialization');
          clear();
        } else if (lastOperationType === 'UPDATE_DOCUMENT') {
          // insert snapshot into the history
          LOG.verbose('insert a new state in undo history', snapshot);
          setState(snapshot);
        } else {
          LOG.verbose("skip the current state as it's not updating the document");
        }
      },
      get,
      api,
    );
  };
