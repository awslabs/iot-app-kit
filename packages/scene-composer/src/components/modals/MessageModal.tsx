import _ from 'lodash';
import { Box, Button, Header, Icon, SpaceBetween, StatusIndicator, TextContent } from '@awsui/components-react';
import React, { Fragment, useContext, useMemo } from 'react';
import { useIntl } from 'react-intl';

import { AWS_USER_GUIDE_DOC_URL } from '../../common/constants';
import { sceneComposerIdContext } from '../../common/sceneComposerIdContext';
import { useStore } from '../../store';
import { DisplayMessageCategory } from '../../store/internalInterfaces';
import { evalStringTemplate } from '../../utils/stringUtils';
import CenteredContainer from '../CenteredContainer';

function mapMessageCategoryToModalTitle(category: DisplayMessageCategory) {
  const intl = useIntl();
  switch (category) {
    case DisplayMessageCategory.Error:
      return intl.formatMessage({ defaultMessage: 'Error - unable to render the scene', description: 'Error message' });
    case DisplayMessageCategory.Warning:
      return intl.formatMessage({ defaultMessage: 'Warning', description: 'Warning message' });
    default:
      return intl.formatMessage({ defaultMessage: 'Notice', description: 'Notice message' });
  }
}

const MessageModal = () => {
  const sceneComposerId = useContext(sceneComposerIdContext);
  const messages = useStore(sceneComposerId)((state) => state.getMessages());
  const clearMessages = useStore(sceneComposerId)((state) => state.clearMessages);
  const intl = useIntl();

  const containsError = messages.find((m) => m.category === DisplayMessageCategory.Error);
  const containsWarning = messages.find((m) => m.category === DisplayMessageCategory.Warning);

  const title = containsError
    ? mapMessageCategoryToModalTitle(DisplayMessageCategory.Error)
    : containsWarning
    ? mapMessageCategoryToModalTitle(DisplayMessageCategory.Warning)
    : mapMessageCategoryToModalTitle(DisplayMessageCategory.Info);
  const allowDismiss = !containsError;
  const showUserGuideLink = containsError || containsWarning;

  const messageContent = useMemo(() => {
    const groupedMessages = _.groupBy(messages, (m) => m.category);
    return [DisplayMessageCategory.Error, DisplayMessageCategory.Warning, DisplayMessageCategory.Info]
      .map((category) => {
        if (category in groupedMessages) {
          const messages = groupedMessages[category];
          if (messages.length === 0) {
            return null;
          }
          return (
            <Fragment key={category}>
              {messages.map((m, index) => (
                <StatusIndicator key={index} type={category.toLocaleLowerCase() as any}>
                  {evalStringTemplate(m.messageText, m.params)}
                  {m.params?.cause ? ' Cause - ' + m.params.cause : ''}
                </StatusIndicator>
              ))}
            </Fragment>
          );
        } else {
          return null;
        }
      })
      .filter((m) => m !== null);
  }, [messages]);

  return (
    <CenteredContainer
      header={<Header variant='h2'>{title}</Header>}
      footer={
        allowDismiss && (
          <Box float='right' padding={{ bottom: 's' }}>
            <Button variant='primary' onClick={clearMessages}>
              Ok
            </Button>
          </Box>
        )
      }
    >
      <SpaceBetween size='s'>
        <Box>{messageContent}</Box>

        {showUserGuideLink && (
          <Box>
            <TextContent>
              <p>
                <a href={AWS_USER_GUIDE_DOC_URL}>
                  {' '}
                  {intl.formatMessage({ defaultMessage: 'Learn More', description: 'Hyperlink title' })}{' '}
                  <Icon name='external' size='normal' variant='link' />{' '}
                </a>
              </p>
            </TextContent>
          </Box>
        )}
      </SpaceBetween>
    </CenteredContainer>
  );
};

MessageModal.displayName = 'MessageModal';

export default MessageModal;
