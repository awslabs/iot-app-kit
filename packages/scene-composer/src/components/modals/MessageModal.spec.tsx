import { render } from '@/tests/testing-library';
import { accessStore } from '../../store';
import { DisplayMessageCategory } from '../../store/internalInterfaces';
import MessageModal from './MessageModal';

import Mock = vi.Mock;

vi.mock('../../store', async () => {
  const originalModule = await vi.importActual('../../store');
  return {
    ...originalModule,
    accessStore: vi.fn(),
  };
});

describe('MessageModalSnap', () => {
  const clearMessages = vi.fn();
  const mockUseStore = accessStore as Mock;

  [
    { category: DisplayMessageCategory.Info, messageText: 'Info Test' },
    { category: DisplayMessageCategory.Warning, messageText: 'Warning Test' },
    { category: DisplayMessageCategory.Error, messageText: 'Error Test' },
  ].map((props) => {
    it(`should render correctly with: ${props}`, () => {
      mockUseStore.mockReturnValueOnce(() => [props]);
      mockUseStore.mockReturnValueOnce(() => clearMessages);

      const { container } = render(<MessageModal />);
      expect(container).toMatchSnapshot();
    });
  });
});
