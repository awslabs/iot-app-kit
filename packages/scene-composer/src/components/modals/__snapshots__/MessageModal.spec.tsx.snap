// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MessageModalSnap > should render correctly with: [object Object] 1`] = `
<div>
  <div
    class="sc-aXZVg eWvTss awsui_root_14iqq_v4tpg_103 awsui_variant-default_14iqq_v4tpg_147 awsui_refresh_14iqq_v4tpg_155"
  >
    <div
      class="awsui_content-wrapper_14iqq_v4tpg_222"
      id=":r1:"
    >
      <div
        class="awsui_refresh_14iqq_v4tpg_155 awsui_header_14iqq_v4tpg_263 awsui_with-paddings_14iqq_v4tpg_303"
      >
        <div
          class="awsui_root_2qdw9_1bpgx_99 awsui_root-variant-h2_2qdw9_1bpgx_141 awsui_refresh_2qdw9_1bpgx_138 awsui_root-no-actions_2qdw9_1bpgx_138"
        >
          <div
            class="awsui_main_2qdw9_1bpgx_160 awsui_refresh_2qdw9_1bpgx_138"
          >
            <div
              class="awsui_title_2qdw9_1bpgx_216 awsui_title-variant-h2_2qdw9_1bpgx_228 awsui_refresh_2qdw9_1bpgx_138"
            >
              <h2
                class="awsui_heading_2qdw9_1bpgx_296 awsui_heading-variant-h2_2qdw9_1bpgx_312"
              >
                <span
                  class="awsui_heading-text_2qdw9_1bpgx_327 awsui_heading-text-variant-h2_2qdw9_1bpgx_338"
                  data-analytics-funnel-key="substep-name"
                  id="heading:r2:"
                >
                  Notice
                </span>
              </h2>
            </div>
          </div>
        </div>
      </div>
      <div
        class="awsui_content_14iqq_v4tpg_222 awsui_with-paddings_14iqq_v4tpg_303"
      >
        <div
          class="awsui_root_18582_z63vw_99 awsui_vertical_18582_z63vw_146 awsui_vertical-s_18582_z63vw_158"
        >
          <div
            class="awsui_child_18582_z63vw_103"
          >
            <div
              class="awsui_root_18wu0_1elzt_99 awsui_box_18wu0_1elzt_219 awsui_color-default_18wu0_1elzt_219 awsui_font-size-default_18wu0_1elzt_235 awsui_font-weight-default_18wu0_1elzt_275"
            >
              <span
                class="awsui_root_1cbgc_azp03_153 awsui_status-info_1cbgc_azp03_171"
              >
                <span
                  class="awsui_container_1cbgc_azp03_131 awsui_display-inline-block_1cbgc_azp03_210"
                >
                  <span
                    class="awsui_icon_1cbgc_azp03_103"
                  >
                    <span
                      class="awsui_icon_h11ix_1l964_104 awsui_size-normal-mapped-height_h11ix_1l964_158 awsui_size-normal_h11ix_1l964_154 awsui_variant-normal_h11ix_1l964_230"
                    >
                      <svg
                        aria-hidden="true"
                        focusable="false"
                        viewBox="0 0 16 16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <circle
                          class="stroke-linejoin-round"
                          cx="8"
                          cy="8"
                          r="7"
                        />
                        <path
                          d="M8 11V8H6"
                        />
                        <path
                          class="stroke-linejoin-round"
                          d="M10 11H6"
                        />
                        <path
                          d="M7.99 5H8v.01h-.01z"
                        />
                      </svg>
                    </span>
                  </span>
                  Info Test
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
      <div
        class="awsui_footer_14iqq_v4tpg_421 awsui_with-divider_14iqq_v4tpg_425 awsui_with-paddings_14iqq_v4tpg_303"
      >
        <div
          class="awsui_root_18wu0_1elzt_99 awsui_box_18wu0_1elzt_219 awsui_p-bottom-s_18wu0_1elzt_651 awsui_f-right_18wu0_1elzt_1053 awsui_color-default_18wu0_1elzt_219 awsui_font-size-default_18wu0_1elzt_235 awsui_font-weight-default_18wu0_1elzt_275"
        >
          <button
            class="awsui_button_vjswe_12tek_107 awsui_variant-primary_vjswe_12tek_251"
            data-analytics-funnel-value="button:r3:"
            type="submit"
          >
            <span
              class="awsui_content_vjswe_12tek_103"
            >
              Ok
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MessageModalSnap > should render correctly with: [object Object] 2`] = `
<div>
  <div
    class="sc-aXZVg eWvTss awsui_root_14iqq_v4tpg_103 awsui_variant-default_14iqq_v4tpg_147 awsui_refresh_14iqq_v4tpg_155"
  >
    <div
      class="awsui_content-wrapper_14iqq_v4tpg_222"
      id=":r6:"
    >
      <div
        class="awsui_refresh_14iqq_v4tpg_155 awsui_header_14iqq_v4tpg_263 awsui_with-paddings_14iqq_v4tpg_303"
      >
        <div
          class="awsui_root_2qdw9_1bpgx_99 awsui_root-variant-h2_2qdw9_1bpgx_141 awsui_refresh_2qdw9_1bpgx_138 awsui_root-no-actions_2qdw9_1bpgx_138"
        >
          <div
            class="awsui_main_2qdw9_1bpgx_160 awsui_refresh_2qdw9_1bpgx_138"
          >
            <div
              class="awsui_title_2qdw9_1bpgx_216 awsui_title-variant-h2_2qdw9_1bpgx_228 awsui_refresh_2qdw9_1bpgx_138"
            >
              <h2
                class="awsui_heading_2qdw9_1bpgx_296 awsui_heading-variant-h2_2qdw9_1bpgx_312"
              >
                <span
                  class="awsui_heading-text_2qdw9_1bpgx_327 awsui_heading-text-variant-h2_2qdw9_1bpgx_338"
                  data-analytics-funnel-key="substep-name"
                  id="heading:r7:"
                >
                  Warning
                </span>
              </h2>
            </div>
          </div>
        </div>
      </div>
      <div
        class="awsui_content_14iqq_v4tpg_222 awsui_with-paddings_14iqq_v4tpg_303"
      >
        <div
          class="awsui_root_18582_z63vw_99 awsui_vertical_18582_z63vw_146 awsui_vertical-s_18582_z63vw_158"
        >
          <div
            class="awsui_child_18582_z63vw_103"
          >
            <div
              class="awsui_root_18wu0_1elzt_99 awsui_box_18wu0_1elzt_219 awsui_color-default_18wu0_1elzt_219 awsui_font-size-default_18wu0_1elzt_235 awsui_font-weight-default_18wu0_1elzt_275"
            >
              <span
                class="awsui_root_1cbgc_azp03_153 awsui_status-warning_1cbgc_azp03_165"
              >
                <span
                  class="awsui_container_1cbgc_azp03_131 awsui_display-inline-block_1cbgc_azp03_210"
                >
                  <span
                    class="awsui_icon_1cbgc_azp03_103"
                  >
                    <span
                      class="awsui_icon_h11ix_1l964_104 awsui_size-normal-mapped-height_h11ix_1l964_158 awsui_size-normal_h11ix_1l964_154 awsui_variant-normal_h11ix_1l964_230"
                    >
                      <svg
                        aria-hidden="true"
                        focusable="false"
                        viewBox="0 0 16 16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          class="stroke-linejoin-round"
                          d="m8 1 7 14H1L8 1z"
                        />
                        <path
                          d="M7.99 12H8v.01h-.01zM8 6v4"
                        />
                      </svg>
                    </span>
                  </span>
                  Warning Test
                </span>
              </span>
            </div>
          </div>
          <div
            class="awsui_child_18582_z63vw_103"
          >
            <div
              class="awsui_root_18wu0_1elzt_99 awsui_box_18wu0_1elzt_219 awsui_color-default_18wu0_1elzt_219 awsui_font-size-default_18wu0_1elzt_235 awsui_font-weight-default_18wu0_1elzt_275"
            >
              <div
                class="awsui_text-content_6absk_1rcdq_100"
              >
                <p>
                  <a
                    href="https://docs.aws.amazon.com/iot-twinmaker/latest/guide/what-is-twinmaker.html"
                  >
                     
                    Learn More
                     
                    <span
                      class="awsui_icon_h11ix_1l964_104 awsui_size-normal-mapped-height_h11ix_1l964_158 awsui_size-normal_h11ix_1l964_154 awsui_variant-link_h11ix_1l964_251 awsui_name-external_h11ix_1l964_264"
                    >
                      <svg
                        aria-hidden="true"
                        focusable="false"
                        viewBox="0 0 16 16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          class="stroke-linecap-square"
                          d="M10 2h4v4"
                        />
                        <path
                          d="m6 10 8-8"
                        />
                        <path
                          class="stroke-linejoin-round"
                          d="M14 9.048V14H2V2h5"
                        />
                      </svg>
                    </span>
                     
                  </a>
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="awsui_footer_14iqq_v4tpg_421 awsui_with-divider_14iqq_v4tpg_425 awsui_with-paddings_14iqq_v4tpg_303"
      >
        <div
          class="awsui_root_18wu0_1elzt_99 awsui_box_18wu0_1elzt_219 awsui_p-bottom-s_18wu0_1elzt_651 awsui_f-right_18wu0_1elzt_1053 awsui_color-default_18wu0_1elzt_219 awsui_font-size-default_18wu0_1elzt_235 awsui_font-weight-default_18wu0_1elzt_275"
        >
          <button
            class="awsui_button_vjswe_12tek_107 awsui_variant-primary_vjswe_12tek_251"
            data-analytics-funnel-value="button:r8:"
            type="submit"
          >
            <span
              class="awsui_content_vjswe_12tek_103"
            >
              Ok
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MessageModalSnap > should render correctly with: [object Object] 3`] = `
<div>
  <div
    class="sc-aXZVg eWvTss awsui_root_14iqq_v4tpg_103 awsui_variant-default_14iqq_v4tpg_147 awsui_refresh_14iqq_v4tpg_155"
  >
    <div
      class="awsui_content-wrapper_14iqq_v4tpg_222"
      id=":rb:"
    >
      <div
        class="awsui_refresh_14iqq_v4tpg_155 awsui_header_14iqq_v4tpg_263 awsui_with-paddings_14iqq_v4tpg_303"
      >
        <div
          class="awsui_root_2qdw9_1bpgx_99 awsui_root-variant-h2_2qdw9_1bpgx_141 awsui_refresh_2qdw9_1bpgx_138 awsui_root-no-actions_2qdw9_1bpgx_138"
        >
          <div
            class="awsui_main_2qdw9_1bpgx_160 awsui_refresh_2qdw9_1bpgx_138"
          >
            <div
              class="awsui_title_2qdw9_1bpgx_216 awsui_title-variant-h2_2qdw9_1bpgx_228 awsui_refresh_2qdw9_1bpgx_138"
            >
              <h2
                class="awsui_heading_2qdw9_1bpgx_296 awsui_heading-variant-h2_2qdw9_1bpgx_312"
              >
                <span
                  class="awsui_heading-text_2qdw9_1bpgx_327 awsui_heading-text-variant-h2_2qdw9_1bpgx_338"
                  data-analytics-funnel-key="substep-name"
                  id="heading:rc:"
                >
                  Error - unable to render the scene
                </span>
              </h2>
            </div>
          </div>
        </div>
      </div>
      <div
        class="awsui_content_14iqq_v4tpg_222 awsui_with-paddings_14iqq_v4tpg_303"
      >
        <div
          class="awsui_root_18582_z63vw_99 awsui_vertical_18582_z63vw_146 awsui_vertical-s_18582_z63vw_158"
        >
          <div
            class="awsui_child_18582_z63vw_103"
          >
            <div
              class="awsui_root_18wu0_1elzt_99 awsui_box_18wu0_1elzt_219 awsui_color-default_18wu0_1elzt_219 awsui_font-size-default_18wu0_1elzt_235 awsui_font-weight-default_18wu0_1elzt_275"
            >
              <span
                class="awsui_root_1cbgc_azp03_153 awsui_status-error_1cbgc_azp03_162"
              >
                <span
                  class="awsui_container_1cbgc_azp03_131 awsui_display-inline-block_1cbgc_azp03_210"
                >
                  <span
                    class="awsui_icon_1cbgc_azp03_103"
                  >
                    <span
                      class="awsui_icon_h11ix_1l964_104 awsui_size-normal-mapped-height_h11ix_1l964_158 awsui_size-normal_h11ix_1l964_154 awsui_variant-normal_h11ix_1l964_230"
                    >
                      <svg
                        aria-hidden="true"
                        focusable="false"
                        viewBox="0 0 16 16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <circle
                          class="stroke-linejoin-round"
                          cx="8"
                          cy="8"
                          r="7"
                        />
                        <path
                          d="m10.828 5.172-5.656 5.656M10.828 10.828 5.172 5.172"
                        />
                      </svg>
                    </span>
                  </span>
                  Error Test
                </span>
              </span>
            </div>
          </div>
          <div
            class="awsui_child_18582_z63vw_103"
          >
            <div
              class="awsui_root_18wu0_1elzt_99 awsui_box_18wu0_1elzt_219 awsui_color-default_18wu0_1elzt_219 awsui_font-size-default_18wu0_1elzt_235 awsui_font-weight-default_18wu0_1elzt_275"
            >
              <div
                class="awsui_text-content_6absk_1rcdq_100"
              >
                <p>
                  <a
                    href="https://docs.aws.amazon.com/iot-twinmaker/latest/guide/what-is-twinmaker.html"
                  >
                     
                    Learn More
                     
                    <span
                      class="awsui_icon_h11ix_1l964_104 awsui_size-normal-mapped-height_h11ix_1l964_158 awsui_size-normal_h11ix_1l964_154 awsui_variant-link_h11ix_1l964_251 awsui_name-external_h11ix_1l964_264"
                    >
                      <svg
                        aria-hidden="true"
                        focusable="false"
                        viewBox="0 0 16 16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          class="stroke-linecap-square"
                          d="M10 2h4v4"
                        />
                        <path
                          d="m6 10 8-8"
                        />
                        <path
                          class="stroke-linejoin-round"
                          d="M14 9.048V14H2V2h5"
                        />
                      </svg>
                    </span>
                     
                  </a>
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
